src/storage/hummock_trace/src/collector.rs

```diff
@@ -0,0 +1,408 @@
+pub fn init_collector() {
+    let path = match env::var(LOG_PATH) {
+        Ok(p) => p,
+        Err(_) => DEFAULT_PATH.to_string(),
+    };
+    let path = Path::new(&path);
+
+    if let Some(parent) = path.parent() {
+        if !parent.exists() {
+            create_dir_all(parent).unwrap();
+        }
+    }
+
+    let f = OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .create(true)
+        .open(path)
+        .expect("failed to open log file");
+
+    let writer = BufWriter::with_capacity(WRITER_BUFFER_SIZE, f);
+    let writer = TraceWriterImpl::new_bincode(writer).unwrap();
+    tokio::spawn(GLOBAL_COLLECTOR.run(Box::new(writer)));
+}
+
+/// `GlobalCollector` collects traced hummock operations.
+/// It starts a collector thread and writer thread.
+struct GlobalCollector {
+    tx: Sender<RecordMsg>,
+    rx: Receiver<RecordMsg>,
+}
+
+impl GlobalCollector {
+    fn new() -> Self {
+        let (tx, rx) = unbounded();
+        Self { tx, rx }
+    }
+
+    async fn run(&self, writer: Box<dyn TraceWriter + Send>) {
+        let (writer_tx, writer_rx) = unbounded();
+
+        let writer_handle = tokio::spawn(GlobalCollector::start_writer_worker(writer_rx, writer));
+
+        let rx = self.rx.clone();
+
+        let collect_handle = tokio::spawn(GlobalCollector::start_collect_worker(rx, writer_tx));
+
+        collect_handle
+            .await
+            .expect("failed to stop collector thread");
+        writer_handle.await.expect("failed to stop writer thread");
+    }
+
+    async fn start_collect_worker(rx: Receiver<RecordMsg>, writer_tx: Sender<WriteMsg>) {
+        let mut records = Vec::new();
+        loop {
+            if let Ok(message) = rx.recv_async().await {
+                match message {
+                    RecordMsg::Left(record) => {
+                        records.push(record);
+                    }
+                    RecordMsg::Right(()) => {
+                        writer_tx
+                            .send(WriteMsg::Left(records))
+                            .expect("failed to send write req");
+                        writer_tx
+                            .send(WriteMsg::Right(()))
+                            .expect("failed to kill writer thread");
+                        return;
+                    }
+                }
+            }
+            if !records.is_empty() && !rx.is_empty() {
+                writer_tx
+                    .send(WriteMsg::Left(records))
+                    .expect("failed to send write req");
+                records = Vec::new();
+            }
+        }
+    }
+
+    async fn start_writer_worker(rx: Receiver<WriteMsg>, mut writer: Box<dyn TraceWriter + Send>) {
+        loop {
+            if let Ok(msg) = rx.recv_async().await {
+                match msg {
+                    WriteMsg::Left(records) => {
+                        writer
+                            .write_all(records)
```

Comment:
Can we make the writer method to be async so that we won't block the tokio worker thread on write? You may refer to tokio io.